trin-core/src/portalnet/overlay_service.rs

```diff
@@ -176,96 +241,228 @@ impl OverlayService {
-            self.process_request(request).await;
+        loop {
+            tokio::select! {
+                Some(request) = self.request_rx.recv() => self.process_request(request),
+                Some(response) = self.response_rx.recv() => {
+                    // Look up active request that corresponds to the response.
+                    let optional_active_request = self.active_outgoing_requests.write().remove(&response.request_id);
+                    if let Some(active_request) = optional_active_request {
+                        // Send response to responder if present.
+                        if let Some(responder) = active_request.responder {
+                            let _ = responder.send(response.response.clone());
+                        }
+
+                        // Perform background processing.
+                        match response.response {
+                            Ok(response) => self.process_response(response, active_request.destination),
+                            Err(error) => self.process_request_failure(response.request_id, active_request.destination, error),
+                        }
+                    } else {
+                        warn!("No active request with id {} for response", response.request_id);
+                    }
+                }
+                Some(Ok(node_id)) = self.peers_to_ping.next() => {
+                    // If the node is in the routing table, then ping and re-queue the node.
+                    let key = kbucket::Key::from(node_id);
+                    if let kbucket::Entry::Present(ref mut entry, _) = self.kbuckets.write().entry(&key) {
+                        self.ping_node(&entry.value().enr());
+                        self.peers_to_ping.insert(node_id);
+                    }
+                }
+                _ = OverlayService::bucket_maintenance_poll(self.protocol.clone(), &self.kbuckets) => {}
+            }
         }
     }
 
     /// Returns the local ENR of the node.
-    async fn local_enr(&self) -> Enr {
+    fn local_enr(&self) -> Enr {
         self.discovery.discv5.local_enr()
     }
 
     /// Returns the data radius of the node.
-    async fn data_radius(&self) -> U256 {
+    fn data_radius(&self) -> U256 {
         *self.data_radius
     }
 
+    /// Maintains the routing table.
+    ///
+    /// Consumes previously applied pending entries from the `KBucketsTable`. An `AppliedPending`
+    /// result is recorded when a pending bucket entry replaces a disconnected entry in the
+    /// respective bucket.
+    async fn bucket_maintenance_poll(
+        protocol: ProtocolId,
+        kbuckets: &Arc<RwLock<KBucketsTable<NodeId, Node>>>,
+    ) {
+        future::poll_fn(move |_cx| {
+            // Drain applied pending entries from the routing table.
+            if let Some(entry) = kbuckets.write().take_applied_pending() {
+                debug!(
+                    "[{:?}] Node {:?} inserted and node {:?} evicted",
```

Comment:
So I take it you ran it and checked how often this debug log was popping up? I worry a bit about how busy it will get when we're connected to a lot of nodes. That's not only a logging concern, it's also a concern about how often we'll run this `take_applied_pending()`. I only have a surface understanding of how the tokio select works, and how that interacts with the `poll_fn` here.

Reply:
This log only pops up when there is an "applied pending entry," and this happens when the following occurs:

- A bucket fills up
- We attempt to insert a new node but the bucket is full, so the node goes is inserted into the "pending" slot
- We attempt to ping the least recently contacted node in that full bucket
- That node fails to respond
- We evict the node who failed to respond and "apply the pending entry" to insert the node from the pending slot into the bucket

I don't suspect that this will occur too frequently, assuming that nodes reach some "steady state" where their routing tables stay relatively stable.

I don't have anything quantitative with respect to how often we will run `take_applied_pending`. From what I recall from my own manual testing with additional debug logs, the function is not called infinitely many times, but only at times where there is work on at least one of the other `select` branches (e.g. a new request).